src/lib/reasoners/fun_sat.ml

```diff
@@ -177,6 +177,12 @@ module Make (Th : Theory.S) = struct
     unit_facts_cache : (E.gformula * Ex.t) ME.t ref;
     last_saved_model : Models.t Lazy.t option ref;
     unknown_reason : Sat_solver_sig.unknown_reason option;
+
+    declare_top : Id.typed list ref;
```

Comment:
It was my first attempt of course for sake of performance but it doesn't work. It seems we need to share this field between several environment so we cannot avoid the double indirection with current implementation.

Reply:
Hm, that may sound like a bug? We do create new environments (copies) all the time, but the old ones should no longer be used (we overwrite the fields in `Fun_sat_frontend`). The `pop` issue should also no longer be relevant, because we can write to the mutable field here. Can you elaborate on what "it doesn't work" means?

Reply:
I will check what's happen. When I tried using a mutable field instead of a reference, FunSAT keeps declarations of popped assertion level. Then I switch to a reference (I only did this modification) and it magically worked.

Reply:
Note that with a mutable field you need to set `acc.declare_top <- Stack.pop acc.declare_tail` not `env.declare_top <- Stack.pop env.declare_tail` in the `pop` function (which I'd guess is the issue given what you describe).